edgeless_api/src/node_registration.rs

```diff
@@ -1,19 +1,57 @@
 // SPDX-FileCopyrightText: © 2024 Technical University of Munich, Chair of Connected Mobility
 // SPDX-License-Identifier: MIT
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct ResourceProviderSpecification {
     pub provider_id: String,
     pub class_type: String,
     pub outputs: Vec<String>,
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct NodeCapabilities {
+    // Number of (actual or virtual) CPUs associated with the edge node.
+    pub num_cpus: u32,
+    // Name of the CPU model.
+    pub model_name_cpu: String,
+    // Clock frequency of the CPU, in BogoMIPS.
+    pub clock_freq_cpu: f32,
+    // Number of cores for each CPU.
+    pub num_cores: u32,
+    // Size of memory available to applications running on the edge node, in MB.
+    pub mem_size: u32,
+}
+
+impl NodeCapabilities {
+    pub fn default() -> Self {
```

Comment:
Those defaults will require matching on inner fields,
As this is used above, you always attach the default if no settings are present, so the logic has to check for fields that don't make sense semantically (e.g., zero CPUs) to determine it is just not set.

Reply:
Yes, indeed.

The rationale for this choice is: based on experience, it is difficult to predict which fields will make sense or will get a meaningful value for all the architectures, especially since this kind of structures tend to grow over time.

So, it will be always a responsibility of who is using the values to translate the entire set of capabilities, some of which may be missing (= default value).

I could have used `Option<>` of course, do you think that would be better?

Reply:
I would go for Option, or, even better, a custom enum similar to { NotSet, Value(value), Infinite } to reduce the amount of guesswork and encode that in the type-system.

Reply:
OK, all this makes a lot of sense. Let's discuss it in the live document to have a broader audience.